tools/oprt2/pkg/commandrunner/runner.go

```diff
@@ -0,0 +1,171 @@
+		r.logger.DebugContext(ctx, "running hook cleanup", "hook", hook.Name())
+		errs = append(errs, hook.Cleanup(ctx))
+	}
+
+	return errors.Join(errs...)
+}
+
+// Setup runs setup hooks. This will return after the first hook failure, or all hooks succeed.
+func (r *Runner) Setup(ctx context.Context) error {
+	for _, hook := range r.hooks {
+		r.logger.DebugContext(ctx, "running hook setup", "hook", hook.Name())
+		if err := hook.Setup(ctx); err != nil {
+			return fmt.Errorf("hook %q setup failed: %w", hook.Name(), err)
+		}
+	}
+
+	r.setupRan = true
+	return nil
+}
+
+// Run executes the provided command. If setup hooks have not been successfully called, they
+// will be ran prior to executing the command.
+// Cleanup hooks will not be called explicitly. The caller should `Close` the runner after
+// all associated command have been ran.
+func (r *Runner) Run(ctx context.Context, name string, args ...string) error {
+	// Ensure setup has ran successfully at least once
+	if !r.setupRan {
+		if err := r.Setup(ctx); err != nil {
+			return fmt.Errorf("setup failed: %w", err)
+		}
+	}
+
+	// Allow hooks to modify the command or arguments
+	for _, hook := range r.hooks {
+		r.logger.DebugContext(ctx, "running hook pre-command", "hook", hook.Name())
+		if err := hook.PreCommand(ctx, &name, &args); err != nil {
+			cmdString := buildCommandDebugString(exec.Command(name, args...))
+			return fmt.Errorf("hook %q pre-command failed for command %q: %w",
+				hook.Name(), cmdString, err)
+		}
+	}
+
+	// Build the command
+	cmd := exec.CommandContext(ctx, name, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+
+	// Allow hooks to modify the command (add env vars, args, etc.)
+	for _, hook := range r.hooks {
+		r.logger.DebugContext(ctx, "running command hook", "hook", hook.Name())
+		if err := hook.Command(ctx, cmd); err != nil {
+			return fmt.Errorf("hook %q command failed on command %q: %w", buildCommandDebugString(cmd),
+				hook.Name(), err)
+		}
+	}
+
+	cmdDebugString := buildCommandDebugString(cmd)
+	r.logger.DebugContext(ctx, "running command", "command", cmdDebugString)
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("command %q failed: %w", cmdDebugString, err)
+	}
+
+	return nil
+}
+
+// buildCommandDebugString builds a textual version of cmd for debug logging.
+// This should never be directly executed.
+func buildCommandDebugString(cmd *exec.Cmd) string {
+	if cmd == nil {
+		return ""
+	}
+
+	// This is the same as `cmd.String()` but it includes env vars
+	// Example: VAR1=val1 VAR2=val2 /path/to/command --with args
+	commandTextBuilder := new(strings.Builder)
+	for _, envVar := range cmd.Env {
+		// The current implementation will never return a a non-nil error, and
+		// handling one would just complicate this logic without much added value
+		_, _ = commandTextBuilder.WriteString(envVar)
+		_, _ = commandTextBuilder.WriteRune(' ')
+	}
+	commandTextBuilder.WriteString(cmd.String())
```

Comment:
Wouldn't `strings.Join(append(cmd.Env, cmd.String()), " ")` (or something like that, I might've messed up syntax a bit but you get the idea) be simpler than this?